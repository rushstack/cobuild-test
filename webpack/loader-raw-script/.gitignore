node_modules
.DS_Store
npm*.log
coverage
temp
lib
dist
