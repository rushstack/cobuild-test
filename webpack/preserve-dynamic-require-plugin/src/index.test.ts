jest.disableAutomock();
import webpack from 'webpack';

import { PreserveDynamicRequireWebpackPlugin } from './index';

jest.setTimeout(1e9);

describe(PreserveDynamicRequireWebpackPlugin.name, () => {
  it('Preserves dynamic require usage', async () => {
    const sources: Map<string, string> = new Map();
    sources.set('/package.json', JSON.stringify({}));
    sources.set('/file.js', 'require(process.env.SOME_PATH);');

    const compiler: webpack.Compiler = webpack({
      entry: {
        main: '/file.js'
      },
      output: {
        path: '/release',
        filename: '[name].js'
      },
      context: '/',
      mode: 'none',
      plugins: [new PreserveDynamicRequireWebpackPlugin()]
    });

    const inputFs: typeof compiler.inputFileSystem = {
      readFile(path: string, cb: (err?: NodeJS.ErrnoException | undefined, content?: string) => void): void {
        cb(undefined, sources.get(path));
      },
      readdir(path: string, cb: (err?: NodeJS.ErrnoException | undefined, files?: string[]) => void): void {
        cb(
          undefined,
          Array.from(sources.keys(), (key: string) => key.slice(1))
        );
      },
      readlink(path: string, cb: (err?: NodeJS.ErrnoException | undefined, dest?: string) => void): void {
        cb();
      },
      stat(path: string, cb: (err?: NodeJS.ErrnoException, stat?: unknown) => void): void {
        if (sources.has(path)) {
          return cb(undefined, {
            isFile() {
              return true;
            },
            isDirectory() {
              return false;
            }
          });
        } else if (path === '/') {
          return cb(undefined, {
            isFile() {
              return false;
            },
            isDirectory() {
              return true;
            }
          });
        }
        cb(new Error(`Unexpected stat call for ${path}`));
      }
    } as unknown as typeof compiler.inputFileSystem;

    const results: Map<string, string> = new Map();

    const outputFs: typeof compiler.outputFileSystem = {
      mkdir(path: string, cb: (err?: NodeJS.ErrnoException) => void): void {
        cb();
      },
      stat(path: string, cb: (err?: NodeJS.ErrnoException) => void): void {
        const err: NodeJS.ErrnoException = new Error(`No such file`);
        err.code = 'ENOENT';
        cb(err);
      },
      writeFile(path: string, content: Buffer, cb: (err?: NodeJS.ErrnoException) => void): void {
        results.set(path, content.toString('utf8'));
        cb();
      }
    } as unknown as typeof compiler.outputFileSystem;

    compiler.inputFileSystem = inputFs;
    compiler.outputFileSystem = outputFs;

    await new Promise<void>((resolve: () => void, reject: (err: Error) => void) =>
      compiler.run((err?: Error | null) => {
        if (err) {
          return reject(err);
        }
        resolve();
      })
    );

    expect(results).toMatchSnapshot();
  });
});
