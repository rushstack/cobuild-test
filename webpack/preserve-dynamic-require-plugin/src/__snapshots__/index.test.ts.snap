// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreserveDynamicRequireWebpackPlugin Preserves dynamic require usage 1`] = `
Map {
  "/release/main.js" => "/******/ (() => { // webpackBootstrap
var __webpack_exports__ = {};
require(process.env.SOME_PATH);
/******/ })()
;",
}
`;
